docs/architecture/adr-011-light-client-crates-extraction.md

```diff
@@ -0,0 +1,292 @@
+* `AnyClientState` implements `ClientState`.
+* `AnyConsensusState` implements `ConsensusState`.
+* `AnyHeader` implements `Header`.
+* `AnyMisbehaviour` implements `Misbehaviour`.
+
+In order to replace the `Any*` enums with trait objects, these light client traits must be object safe. This
+essentially means that these traits cannot have a `Self: Sized` requirement and their methods cannot have type
+parameters & cannot use `Self`. These restrictions, their implications and possible workarounds are documented below.
+
+#### Traits that require `Self: Sized` cannot be used as supertraits
+
+Common traits that the light client traits depend on that have the `Self: Sized` requirement
+are `core::convert::{Into<T>, From<T>, ...}`, `tendermint_proto::Protobuf`, `Clone`, etc.
+
+An easy workaround is to add these methods directly to the trait, for e.g. we add an `encode_vec()` method to
+the `Header` trait directly.
+
+```rust
+pub trait ClientState {
+    /// Encode to canonical binary representation
+    fn encode_vec(&self) -> Vec<u8>;
+
+    /* ... */
+}
+/// instead of ->
+/// pub trait ClientState: Protobuf<Any> { /* ... */}
+```
+
+#### `Clone` cannot be a supertrait and cannot be derived for types containing boxed trait objects
+
+The following can be made to work using the `dyn-clone` crate ->
+
+```rust
+#[derive(Clone)]
+pub struct Result {
+    pub client_id: ClientId,
+    pub client_state: Box<dyn ClientState>,
+    pub consensus_state: Box<dyn ConsensusState>,
+}
+```
+
+```rust
+pub trait ClientState: dyn_clone::DynClone { /* ... */ }
+
+dyn_clone::clone_trait_object!(ClientState);
+```
+
+`dyn-clone` also provides a `clone_box()` function that can be used to get a `Box<T>` from a `&T`.
+
+#### Deriving `serde::{Serialize, Deserialize}` for trait objects
+
+Replacing the `AnyHeader` with a `Box<dyn Header>` wouldn't work in the code below because `Deserialize`
+requires `Self: Sized` and `Serialize` uses generic types.
+
+```rust
+#[derive(Deserialize, Serialize)]
+pub struct UpdateClient {
+    pub common: Attributes,
+    pub header: Option<AnyHeader>,
+}
+```
+
+This can be solved using the `ibc_proto::google::protobuf::Any` type instead and having the light client traits provide
+an encoding to `Any` ->
+
+```rust
+use ibc_proto::google::protobuf::Any;
+
+pub trait Header {
+    /// Encode to canonical protobuf representation
+    fn encode_any(&self) -> Any;
+
+    /* ... */
+}
+```
+
+Thankfully, the core modules code doesn't use the `serde` derivations except in logs and errors. Host and light-client
+implementations can optionally choose to downcast to the concrete type and use it's `serde` derivations directly (if
+available).
+
+#### Light client traits cannot have constructors
+
+This restriction comes from the fact that trait methods of object safe traits cannot return `Self`.  
+However, we would need a ctor to be able to create a `ClientState` and `ConsensusState` in the `create_client` handler.
```

Comment:
Not sure what a "ctor" is. Or is that a typo?

Reply:
`ctor` is shorthand for constructor but I decided to not use it. :) Fixed in https://github.com/informalsystems/ibc-rs/pull/2356/commits/5d2495c3e645ca215f9c76f34eff2080c2d0283d